nifi-nar-bundles/nifi-deeplearning4j-bundle/nifi-deeplearning4j-processors/src/main/java/org/apache/nifi/processors/deeplearning4j/DeepLearning4JPredictor.java

```diff
@@ -0,0 +1,218 @@
+           for (int row = 0; row < inputRecords.length; row++) {
+                INDArray result = results.getRow(row);
+                partitionedResults[row] = Nd4j.toFlattened(result).toDoubleVector();
+           }
+
+           String jsonResult = gson.toJson(partitionedResults);
+           int [] shape = results.shape();
+           String jsonShape = gson.toJson(Arrays.copyOfRange(shape, 1, shape.length));
+
+           if ( getLogger().isDebugEnabled() ) {
+                getLogger().debug("Prediction for inputRecords {}, dims {}, results {}, result.shape {}, partitionedResults {}, jsonResult {}, shape {}, jsonShape {}",
+                        new Object[] {inputRecords, dimensions, results, Arrays.toString(results.shape()), partitionedResults, jsonResult, shape, jsonShape});
+           }
+
+           try (ByteArrayInputStream bais = new ByteArrayInputStream(jsonResult.getBytes(charset))) {
+               flowFile = session.importFrom(bais, flowFile);
+           }
+
+           session.putAttribute(flowFile, DEEPLEARNING4J_OUTPUT_SHAPE, jsonShape);
+
+           final long endTimeMillis = System.currentTimeMillis();
+
+           session.transfer(flowFile, REL_SUCCESS);
+
+           session.getProvenanceReporter().send(flowFile, makeProvenanceUrl(context),
+                    (endTimeMillis - startTimeMillis));
+        } catch (Exception exception) {
+            flowFile = populateErrorAttributes(session, flowFile, exception.getMessage());
+                getLogger().error("Failed to process data due to {} for input {}",
+                        new Object[]{exception.getLocalizedMessage(), input}, exception);
+                session.transfer(flowFile, REL_FAILURE);
+            context.yield();
+        }
+    }
+
+    protected String getFlowFileContents(final ProcessSession session, Charset charset, FlowFile incomingFlowFile)
+            throws IOException {
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+            session.exportTo(incomingFlowFile, baos);
+            return new String(baos.toByteArray(), charset);
+        }
+    }
+
+    protected int [] getInputDimensions(final ProcessContext context, Charset charset, FlowFile flowFile, String separator)
+            throws IOException {
+        String values = context.getProperty(RECORD_DIMENSIONS).evaluateAttributeExpressions(flowFile).getValue();
+        return Arrays.stream(
+                values.split(separator))
+                    .mapToInt(val -> Integer.parseInt(val)).toArray();
+    }
+
+    protected String makeProvenanceUrl(final ProcessContext context) {
+        return new StringBuilder("deeplearning4j://")
+            .append(context.getProperty(MODEL_FILE).evaluateAttributeExpressions().getValue()).toString();
+    }
+
+    protected FlowFile populateErrorAttributes(final ProcessSession session, FlowFile flowFile,
+            String message) {
+        Map<String,String> attributes = new HashMap<>();
```

Comment:
No need to create a HashMap here -- can just use `flowFile = session.putAttribute(flowFile, DEEPLEARNING4J_ERROR_MESSAGE, String.valueOf(message);` -- and if just that single line would probably eliminate the method all together and do inline.

Reply:
Corrected.